app/controllers/items/search_code.js

```diff
@@ -1,44 +1,112 @@
-import Ember from 'ember';
-import SearchCode from '../search_code';
-import AjaxPromise from 'stock/utils/ajax-promise';
-const { getOwner } = Ember;
+import Ember from "ember";
+import SearchCode from "../search_code";
+import capitalize from "lodash/capitalize";
 
 export default SearchCode.extend({
   item: null,
   inlineDescription: true,
   store: Ember.inject.service(),
+  messageBox: Ember.inject.service(),
+  packageService: Ember.inject.service(),
+  i18n: Ember.inject.service(),
 
-  allPackageTypes: Ember.computed("fetchMoreResult", 'item.isSet', function(){
-    if(this.get('item.isSet')) {
-      return this.get('item.setItem.code').allChildPackagesList();
+  allPackageTypes: Ember.computed("fetchMoreResult", "item.isSet", function() {
+    if (this.get("item.isSet")) {
+      return this.get("item.setItem.code").allChildPackagesList();
     } else {
-      return this.store.query('code', { stock: true });
+      return this.store.query("code", { stock: true });
     }
   }),
 
+  deleteAndAssignNew(packageType) {
+    const item = this.get("item");
+    const detailType = item.get("detailType");
+    const detailId = item.get("detail.id");
+    this.showLoadingSpinner();
+    this.get("packageService")
+      .deletePackage(detailType, detailId)
+      .then(() => {
+        if (this.isSubformPackage(packageType)) {
+          this.assignNew(packageType);
+        } else {
+          this.assignNew(packageType, { deleteDetailId: true });
+        }
+      });
+  },
```

Comment:
You can save a lot of lines by using async functions. This applies generally, just using this function as an example.
(for the spinner, there is an async mixin you can use )

```suggestion
async deleteAndAssignNew(packageType) {
    const detailType = item.get("item.detailType");
    const detailId = item.get("item.detailId");

    await this.runTask(this.get("packageService").deletePackage(detailType, detailId));
    this.assignNew(packageType, { deleteDetailId: this.isSubformPackage(packageType) });
},
```